Minor say code refactor, fixes signers sometimes using the wrong verbs over radio

00-Steven opened the pull request:
## About The Pull Request

Alternative title: "Baby's third pr: surprise say code refactor"

While testing sign language I worked out that if the first message sent after toggling sign language involves speaking directly into a radio, it proceeds to use the wrong verb. After several hours of digging through say and telecomms code, I worked out the cause:
the `verb_say` on humans is primarily updated to match tongue data during the verb _selection_ method `say_mod()` in `human_say.dm`, and as a consequence this meant the verb wasn't updated to the correct one until far after it had been copied to a virtualspeaker, leading to the incorrect verb being used for the radio message.

This pr fixes this by refactoring this verb updating behaviour to be in a new method `update_verbs()`, which we then call _before_ attempting to call `radio()` to avoid wrongful copying and thus fixing the bug.

There's theoretically also cases where the previous behaviour would use the wrong verb if one lost or gained a tongue between radio messages, or the wrong verb from those picked from a list for some simple/basic animals, but those are significantly less common.

To note, I gave the new `update_verbs()` method the same arguments as the `say_mod()` method it got split out of, but having gone through all instances of it being used for this I don't feel it actually *needs* those arguments. So I'm unsure whether it should keep them or not. Opened as a draft for now for sanity's sake, and also cause, y'know, it touches say code.
## Why It's Good For The Game

It was _really_ annoying.
## Changelog
:cl:
fix: Signers no longer use the wrong verb when speaking directly into a radio for the first message after toggling sign language.
refactor: Moved the updating of verb variables into a new method which is called earlier in living's say, which should avoid this happening for other things which updated their verbs the same way.
/:cl:

00-Steven committed:
Create update_verbs(), fix related bug

Adds a new method update_verbs() to living_say.dm, as a hook for updating verbs.
Refactors say_mod() in human_say.dm, simple_animal.dm, and basic.dm to instead use update_verbs() as they are updating verb variables rather than selecting them.
Puts this before radio() in say(), fixing a verb mismatch bug from using say_mod() for updating verbs.

MrMelbert commented on code/modules/mob/living/living_say.dm:
```diff
@@ -528,6 +532,14 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
+ * This is primarily a hook for inheritors,
+ * like human_say.dm's tongue-based verb updates.
+ */
+/mob/living/proc/update_verbs(input, list/message_mods = list())
```
`update_verbs` is a bad name for this given the existence of, well, `/verb`s. 
Perhaps `update_say_verb`.

00-Steven replied:
Fair point, I chose such purely cause there are other `verb_X`s, and making this specifically for `verb_say` in name felt off.

MrMelbert commented on code/modules/mob/living/living_say.dm:
```diff
@@ -211,6 +211,10 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
 		if(succumbed)
 			succumb()
 		return
+	
+	//Updates the verb variables of this living before sending any messages
+	//This avoids copying the wrong verbs to the radio's virtualspeaker
+	update_verbs(message, message_mods)
```
Though it seems a little... excessive to call this every time regardless of if `verb_say` has changed? 

Why can we not just update `verb_say` of the mob when it is necessary to - ie when inserting / removing a tongue, or when the tongue's temp say mod changes, or when certain traits are added or remove?

00-Steven replied:
Ah, the previous method was already called every time a message is _composed_, regardless of if `verb_say` had actually changed. Most commonly I believe this would be called from `hear() > compose_message() > say_quote() > say_mod()`.
So given I felt running it whenever you say something is still broadly saner than running it for anyone who hears your message, I didn't think it'd be a significant issue for this pr.

That aside, it _is_ probably a better idea.
I believe for `human.dm` instances it only matters when the tongue gets added/removed or the `temp_say_mod` gets changed by sign language (the only thing which changes it). So looking into that deeper, I think it'd be fine to call it on `Insert()` and `Remove()` in `_tongue.dm` and on the sign language component when it would change the speaker's `verb_X`s. `Insert()` and `Remove()` should also be after the `COMSIG_CARBON_GAIN/LOSE_ORGAN` signals and so update it after sign language's changes.

Now for the caveats, there's the basic/simple animals which used it to update `verb_say` like this:
```byond
/mob/living/simple_animal/say_mod(input, list/message_mods = list())
	if(length(speak_emote))
		verb_say = pick(speak_emote)
	return ..()
```
Which would have the same issues, though significantly less common. But there's also no easy "update it only on changes" cause it's _intended_ to change every say attempt.
So maybe updating it at the end of `say()` if it returns `TRUE` in `basic.dm` and `simple_animal.dm` would be fine?
Like say:
```byond
/mob/living/simple_animal/say(pile of arguments)
	. = ..()
	if (.)
		update_say_verb()
	return .
```

I'll be testing such when I have more time tomorrow.

00-Steven replied:
Had some circumstances so it took me a while to get back to you.

On the actual topic, hmn, so that sounds good, and yes virtualspeaker is where the original issue lies, but there's an immediate issue I see with this implementation regarding simple_animal/basic.
Because they use `pick()`, `get_default_say_verb()` will pick a _random_ say verb each time and thus the virtualspeaker verb mismatch issue wouldn't be solved for them as it might return a different verb for the virtualspeaker initialization than for the initial message.
![image](https://github.com/tgstation/tgstation/assets/42909981/4832a035-bfd1-41da-acbe-a8b5bbbbbd00)
Even causing this to be a new issue regarding simple_mob/basic instances speaking into radios/somesuch at all
Besides that I feel it'd be broadly better, there's only a few more `verb_say` reading occurrences we'd have to change, and also tested it and it works fine. I'll have to sit down and think of a good solution regarding the simple_mob/basic issue in a bit though.

On the final bit, do we even _have_ access modifiers? I can't find any in the reference nor have i seen any. Do we have linting for it? Regardless I feel it's probably important we update `code/modules/mob/say_readme.md` to match this change.,

00-Steven replied:
One of the `verb_say` reading occurrences is in `/obj/machinery/ai_voicechanger`, which reads its owner AI's `verb_say`, and upon closer inspection seems to do so for storing it before replacing it with a different one.
Under the proposed system, I feel we'd need to gut the current logic and instead have the `ai_voicechanger` instance store only the verbs it's changing into while adding the verb selection logic to the owner AI's `say_mod()` and `get_default_say_verb()`.

00-Steven replied:
With irl circumstances almost over, just one-two more days to go and I can get back to this.
I care about it cause under the proposed logic it'd it'd be _more_ broken than before.
Before it would only update `verb_say` after the speak-into-radio part and thus the virtualspeaker created before then would have the wrong `verb_say`, but with the proposed logic creating a virtualspeaker for a basic/simple animal is guaranteed to pick a random say verb regardless- so including when they're heard over radios _at all_. Hell, I think it'd even cause the say verb to randomly be different for different listeners regardless of radio.

Y'know, I feel like this'd all be much less of a hassle if the verb used was included in the message data rather than retrieved from the speaker with every listener, so there is one definitive point where the verb is decided on and then it just carries over. Entirely removing the need to worry about whether the say verb we're accessing is still the right say verb in the first place.
There's probably some catches to this, like talking into a radio directly making you whisper, but I'll have to wait with looking deeper into the viability of such for a few more days.

00-Steven committed:
Implement reviewer suggestions, much still to do

Makes say_mod() a pure proc.
Replaces update_verbs() with get_default_say_verb(), which is now in movable atoms rather than living.
Replaces most uses of reading verb_say with get_default_say_verb().
Introduces a novel bug to basic and simple animals.

00-Steven committed:
Resolves comms verb consistency issues

Calls and adds `say_mod()` to message mods as `SAY_MOD_VERB`, setting a definitive point for the verb of a message to be declared and thus letting us solve all verb consistency issues with less hassle.
`say_quote()` does still call `say_mod()` directly if no `SAY_MOD_VERB` message mod was set as a failsafe.
Similarly `say()` in `living_say.dm` does call `say_mod()` and change `SAY_MOD_VERB` again after talking directly into a radio, as it needs to make the original verb a whisper. `say_mod()` is called again instead of directly setting it to `verb_whisper` as it checks and returns whispering first anyway while in the future other inheritors of `living_say.dm` might have a different priority.

00-Steven committed:
Resolves comms microphone whisper consistency issues

Explicitly whispering directly into a radio device would star the message, according to the code "as radios don't pick up whispers very well".
However, whispering across the room from any radio device with the microphone turned on picked up your whispers just fine.
This fixes that.

00-Steven commented:
Alright so, explaining the changes since the start of this pr.
Originally I was trying to fix this obscure verb mismatch bug where using comms directly after toggling sign language would use the wrong verb for your message.
While working out how, I realized this is because toggling sign language would make you use the "signs" verb by changing a temporary modifier on the speaker's tongue, which in turn would only be updated to be the speaker's `verb_say` when `say_mod()` would be called. This method, however, is only called on the *listener*, meaning it'd only get updated after the `virtualspeaker` for the radio message was created and thus given the old wrong `verb_say`.

Trivially, this was fixed by simply calling `say_mod()` before sending a radio message, but while better this was obviously janky and with Melbert's help this was shifted to a saner logic. First making it a separate function, then only updating the verbs when *absolutely* necessary, and finally just not trying to change `verb_say` and instead having a function to get the correct say verb taking the place of `verb_say` in `say_mod()` and when setting the `virtualspeaker`'s `verb_say`.

Now, this fixed the signing issue perfectly, but made a few issues more pronounced. Apparently some basic/simple animals would randomly select a say verb to use, which would now be *guaranteed* to be randomly picked for different comms messages- a cat meowing next to two microphone on radio devices might purr in one and meow in the other. Sure it's less of an issue, but it's still an issue, and I'd rather not knowingly introduce more issues into this pile of technical debt thanks.

So I thought, maybe it'd be better to include the verb in the message `message_mods`, so it gets set at one definitive point and gets carried over to whatever derivative message without changing it. And so I did some digging to work that out, and found out there's more comms-based consistency issues here than I previously thought.

1. The original signer toggling comms verb mismatch.
![image](https://github.com/tgstation/tgstation/assets/42909981/74d76a9a-8c5e-4a8e-8ec5-9e03c0d2aa85)

2. The directly adjacent basic/simple animal randomized verb comms verb mismatch.
(This is not normally visible due to talking directly into a radio making you whisper, but noted as the naive fix above made this apparent)

3. While you can sing or yell or exclaim or such over comms, you cannot slur or stammer or incoherently sign or somesuch. This is because these verbs are decided in `living_say.dm` while the `virtualspeaker` used is not an instance of `living`.
![image](https://github.com/tgstation/tgstation/assets/42909981/e92768bf-e271-4804-95f7-b1898be943d4)
When talking over comms directly this is less obvious as it replaces the original verb with a whisper.
![image](https://github.com/tgstation/tgstation/assets/42909981/9cda7111-d109-405f-a5d9-59982b455300)

4. While you can sing or whisper when talking *directly* over comms, radio devices with the microphone setting on do *not* pick this up. This is because they get a filtered list of message mods excluding whisper/sing mode, and so those're excluded from working out which verb should be used.
![image](https://github.com/tgstation/tgstation/assets/42909981/c7a59f4d-5f92-4b28-bacc-915fcb80778d)

5. Explicitly whispering *directly* into into a radio device stars the message, as the radio can't pick it up well, while radio devices with the microphone setting on pick it up perfectly AND as per 4 do not show you whispering either.
![image](https://github.com/tgstation/tgstation/assets/42909981/2805a9e5-a70f-4176-bd79-d3c1e5e30c82)

The current iteration of this pr fixes 1-4 by calling `say_mod()` *before* sending a radio message and adding it to `message_mods` as `SAY_MOD_VERB`, only changing such when need be, and making this carry over properly to derivative radio messages. `say_quote()` similarly now only calls `say_mod()` if no `SAY_MOD_VERB` is found.
Now it feels *off* to use `message_mods` as such, because it's primarily created in `get_message_mods()` by chopping off symbols to work out radio/whispering/singing/such. But hey! It's called `say_mod()`, `message_mods` is modified in other ways too, and more importantly trying to do this without using `message_mods` feels like it'd be more liable to break things than it's worth.

It also fixes 5 by adding the code for starring messages in `radio`'s `Hear()` before the device propagate it, though I feel I could move this and `living_say.dm`'s starring of whispered radio messages to the same spot in radio code.
There's a few more adjustments like that I'm aware of which I'll be trying to look into in a bit, if nothing kneecaps my free time.

00-Steven commented on code/modules/mob/living/living_say.dm:
```diff
@@ -312,7 +316,7 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
 	if(speaker != src)
 		if(!radio_freq) //These checks have to be separate, else people talking on the radio will make "You can't hear yourself!" appear when hearing people over the radio while deaf.
-			deaf_message = "[span_name("[speaker]")] [speaker.verb_say] something but you cannot hear [speaker.p_them()]."
+			deaf_message = "[span_name("[speaker]")] [speaker.get_default_say_verb()] something but you cannot hear [speaker.p_them()]."
```
To note about this line, previously using `[speaker.verb_say]` here would make it use either the default `verb_say` of the speaker, if it's a human gurgling if it doesn't have a tongue or whatever the tongue `say_mod` is if it does have one, or if it's a basic/simple animal whichever random verb from the list `verb_say` is currently updated to.

Using `[speaker.get_default_say_verb()]` is the closest to previous functionality, and though it means the verb it picks for basic/simple animals will be a different from the original message it doesn't matter much because *you're deaf* and can't hear the original message nor any derivative comms messages anyway.

However, I want to pitch two options here:
If we were to keep `[speaker.verb_say]` now, it now would pretty much always be the default `verb_say` of the speaker, regardless of tongue `say_mod` or lack of tongue or verb randomization, which might be desirable.
We could also replace it by `[message_mods[SAY_MOD_VERB] || speaker.say_mod(message, message_mods)]`, which should maintain previous functionality in addition to allowing deaf `living` instances to see the difference between someone speaking/yelling/singing/whispering/such at them.

00-Steven committed:
Add SAY_MOD_VERB to message mods in say.dm say()

Previously relied on the failsafe using the previous methodology in `say_quote()`, now consistent with updated behaviour

00-Steven committed:
radio.dm Hear() now forwards whispering/singing mods

Includes `WHISPER_MODE` and `MODE_SING` in the filtered message mods sent to `talk_into()`.

00-Steven committed:
Consolidate radio whisper message starring

Moves the starring effect you get when whispering across radios from `living_say.dm` `say()` and `radio.dm` `Hear()` to just `radio.dm` `talk_into()`.

00-Steven commented on code/modules/mob/living/living_say.dm:
```diff
@@ -212,14 +212,13 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
 	//This is before anything that sends say a radio message, and after all important message type modifications, so you can scumb in alien chat or something
 	if(saymode && !saymode.handle_message(src, message, language))
 		return
 	var/radio_message = message
```
`var/radio_message = message`
I feel this could be removed with the directly following message edit being moved, but I'm unsure given ✨ byond ✨.

Ghommie replied:
```suggestion
```

00-Steven commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -354,6 +360,13 @@
+	if (message_mods[WHISPER_MODE])
+		filtered_mods[WHISPER_MODE] = message_mods[WHISPER_MODE]
+	if (message_mods[SAY_MOD_VERB])
+		filtered_mods[SAY_MOD_VERB] = message_mods[SAY_MOD_VERB]
```
There has to be a way to do this saner.
Aside from that, is there even a reason we are filtering the input message mods?
I believe talking directly into a radio forwards them all into `talk_into()` anyway.

00-Steven commented:
...Yeah I'll just try undraft it in a bit then.

00-Steven commented:
> ...Yeah I'll just try undraft it in a bit then.
Words spoken by someone without a working sense of time.

Ghommie commented on code/modules/mob/living/living_say.dm:
```diff
@@ -212,14 +212,13 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
-		radio_message = stars(radio_message)
-		spans |= SPAN_ITALICS
 
 	var/radio_return = radio(radio_message, message_mods, spans, language)//roughly 27% of living/say()'s total cost
```
```suggestion
	var/radio_return = radio(message, message_mods, spans, language)//roughly 27% of living/say()'s total cost
```

Ghommie committed:
Apply suggestions from code review
